video: stm32_dcmi: Snapshot mode

KurtE opened the pull request:
Add the ability to run cameras in a logical
snapshot mode instead of always running
in video mode.  In particular, the camera
starts when you ask for an image to be dequeued
and stops when it receives an image.

There are three ways to setup to run in this mode:

a) define a max of 1 buffer:
CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=1

b) Tell it to do so in the overlay by giving the DCMI object the property: snapshot-mode

c) At run time:  I added a video control: VIDEO_CID_SNAPSHOT_MODE Which can have the values 0 - normal, 1 - snapshot

As mentioned in snapshot mode, the HAL DMA is stopped after each image and restarted when the user asks for another image.  I added a timeout default to 1 second, that handles the case where sometimes the HAL is silently not successful in retrieving an image.

Note: This is the second half of what used to be #93797

KurtE committed:
video: stm32_dcmi: Snapshot mode

Add the ability to run cameras in a logical
snapshot mode instead of always running
in video mode.  In particular, the camera
starts when you ask for an image to be dequeued
and stops when it receives an image.

There are two ways to setup to run in this mode:

a) Tell it to do so in the overlay by giving the DCMI object
the property: snapshot-mode

b) At run time:  I added a video control: VIDEO_CID_SNAPSHOT_MODE
Which can have the values
    0 - normal - continuous video mode
    1 - snapshot - snapshot mode enabled.

As mentioned in the snapshot modes, the HAL DMA is stopped after each
image and restarted when the user asks for another image.  I added a
timeout default to 1 second, that handles the case where sometimes
the HAL is silently not successful in retrieving an image.

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

KurtE commented:
Update: 

I added 2nd mode to snapshot to experiment with starting and stopping the video streams along with 
the HAL start/stop.

Wondering if it might help or hurt on startup speed.  That is in Snapshot mode:
HAL_DCMI_START_DMA with the DCMI_MODE_SNAPSHOT, the code is supposed to align itself to the start of
a frame before it starts to return data.   Now if it is free running maybe it has to wait through almost a full 
frame before it sees the start of next frame.  Maybe if it is started up fresh, maybe it starts a a new frame...

Also was wondering if it might help reduce tearing when it gets an error.  So far It looks like I at times it still
gets into tearing images if a DMA error happens.   I think I am getting fewer of those errors in this mode,
but still at times showing split images and often takes until next error to maybe recover.

Note: I have not yet adding the stop/start in the DMA recovery code will try that as well.

Also wondering if maybe something with the DMA buffer should be cleared when error happens.

etienne-lms commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -314,19 +345,88 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
 	}
 
+	if (HAL_DCMI_Stop(&data->hdcmi) != HAL_OK) {
+		LOG_WRN("Snapshot: HAL_DCMI_Stop FAILED!");
```
Failing to stop DCMI is not an issue here?

etienne-lms commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -338,6 +438,32 @@ static int video_stm32_dcmi_get_caps(const struct device *dev, struct video_caps
+
+	LOG_INF("Set Snapshot: %d\n", value);
+	if ((CONFIG_VIDEO_BUFFER_POOL_NUM_MAX == 1) && (value == 0)) {
+		LOG_DBG("Only one buffer so snapshot mode only");
```
Worth a `LOG_ERR()`?

KurtE replied:
I left it as debug as in many of the cases where one is using snapshot, they really only need/use 1 buffer,
although they can use multiple if they desire

etienne-lms reviewed (commented):
Mainly minor comments.

KurtE commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -37,14 +39,22 @@ struct stream {
 	struct dma_config cfg;
 };
 
+
```
done

KurtE replied:
done

KurtE commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -314,19 +345,88 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
 	struct video_stm32_dcmi_data *data = dev->data;
+	const struct video_stm32_dcmi_config *config = dev->config;
+
+	LOG_DBG("dequeue snapshot: %p %llu\n", data->vbuf, timeout.ticks);
```
done, on several of them... They were printks earlier

KurtE commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -314,19 +345,88 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
 	*vbuf = k_fifo_get(&data->fifo_out, timeout);
 	if (*vbuf == NULL) {
+		uint32_t time_since_start_time =
+			(uint32_t)(k_uptime_get_32() - data->snapshot_start_time);
```
done

KurtE commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -314,19 +345,88 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+			return 0;
+		}
+		data->vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
+		LOG_DBG("\tcamera buf: %p\n", data->vbuf);
```
removed

etienne-lms reviewed (commented):
Do you want to get 2nd commit ("video: stm32_dcmi: snapshot stream stop") also merged or is it a draft as the commit message suggests?

If to merge, I think you should update the commit message. Also prefer have fixes for the 1st commit squashed straight into the 1st commit so that 2nd commit only deal with stream stop mode support.

*(or maybe squash the 2 commits to propose snapshot start/stop modes straight with a single change)*

josuah commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -152,6 +152,12 @@ enum video_colorfx {
+ * Defaults to video mode unless max buffer count is set to 1
+ * 1 - enable snapshot mode
+ */
+#define VIDEO_CID_SNAPSHOT_MODE (VIDEO_CID_BASE + 34)
```
Video controls try to keep compatibility with Linux, and on Linux-side, `_CID_BASE + 34` is already taken:

```
#define V4L2_CID_ROTATE                         (V4L2_CID_BASE+34)
```

It seems something somewhat specific to DCMI. An internal working that affects other aspects of the capture.

In this case, it is possible to introduce a custom CID for it, specific to DCMI, like it is done here for instance:

https://github.com/zephyrproject-rtos/zephyr/pull/96234/files#diff-ec044dc1b7c1624b290dd5eaf094965fb1c6e7124caa174b95682947fc6454de

josuah replied:
Related:

- https://github.com/zephyrproject-rtos/zephyr/issues/104551

KurtE replied:
> Sorry for such a long silence! I am just walking now through the backlog of pull requests after being completely absorbed by work.

Understand... 

> If it is possible to get rid of the `memcpy()` by using snapshot mode, and that it is possible to continuously trigger the next frame, then it sounds like a reasonable thing to only support snapshot mode if the over head is small enough?

I can see several different modes of operation.

**1 buffer** - example Nicla Vision - It has a 5mp camera on it (GC2145), but only main memory available,
With these snapshot with just one buffer works great in that it allows you to get a bigger image

2 or more buffers - that you round robin or the like,  Where as you mention, when you finish a buffer you replace it with a new buffer to receive the next image and hand off the buffer to the caller to be processed. 

Current way - the camera always sees one buffer and does the memcpy. The one case where this can be
nice is if you can allocate a fixed buffer for the camera that is in the Real memory, which then minimizes
the occurrences of DMA errors.

Will take a look

josuah replied:
Sorry for the *very* delayed review on this :(

> 1-buffer/2-buffer

Yes absolutely. This is the way to go.

> Current way (internal SRAM for main frame, external PSRAM

This is a good idea. This would bring us back to have a buffer locally allocated as I think the video stack only supports a single source for buffers.

Then this simplifies the logic as there is always one buffer to store.

Maybe this is less effort for you to schedule this for another PR so that you only have one challenge to solve at a time (if planned at all)? Whichever strategy is your favorite of course.

josuah commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -334,6 +334,12 @@ enum video_camera_orientation {
+ * Defaults to video mode unless max buffer count is set to 1
+ * 1 - enable snapshot mode
+ */
+#define VIDEO_CID_SNAPSHOT_MODE (VIDEO_CID_CAMERA_CLASS_BASE + 36)
```
I think a CID is not bad idea, convenient to swap at runtime.
Though camera IDs are going to be quite standardized.

A lot easier to introduce this exact same text in `include/zephyr/drivers/video/video_stm32_dcmi.h` as `VIDEO_CID_PRIVATE_BASE + 0`, and let user call that custom control.

Then in the future, another driver also needs this, and this can be standardized, but no need to bother about this.

KurtE replied:
Changing now:
```
/** Switch the camera between video mode and snapshot mode.
 * Defaults to video mode unless max buffer count is set to 1
 * 1 - enable snapshot mode
 */
#define VIDEO_CID_SNAPSHOT_MODE (VIDEO_CID_PRIVATE_BASE + 0)
```

Also moved it down below:
```
/**
 * @name Vendor-specific class control IDs
 * @{
 */
#define VIDEO_CID_PRIVATE_BASE 0x08000000
```

Should be in next push

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -270,22 +293,28 @@ static int video_stm32_dcmi_set_stream(const struct device *dev, bool enable,
 	}
 
-	data->vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
+	if (!data->snapshot_mode && (CONFIG_VIDEO_BUFFER_POOL_NUM_MAX != 1)) {
```
I think we face the issue where there is no guarantee that the user will effectively use these buffers for this driver. It could be for a different video device.

So `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX` cannot really be reliably used to know if it is snapshot mode or continuous mode.

Probably safe to trust your own boolean variable `data->snapshot_mode` only.

```suggestion
	if (!data->snapshot_mode) {
```

Or even easier:

```suggestion
	if (!data->ctrls.snapshot_mode.val) {
```

KurtE replied:
I am trying out the: !data->ctrls.snapshot_mode.val
If works, will push up in next push

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -37,22 +39,32 @@ struct stream {
 	struct video_buffer *vbuf;
 	struct stream dma;
+	uint32_t snapshot_start_time;
+	uint8_t snapshot_mode;
```
I think it is possible to remove `snapshot_mode` and use `data->ctrls.snapshot.val` directly instead.

KurtE replied:
As you mentioned, we can probably get rid of this variable, and redirect all the places to use it to 
`!data->ctrls.snapshot_mode.val
`

But need to check where ctrls.snapshot_mode.val is initialized...

josuah replied:
It is initialized here, as part of the video control init. Set to `VIDEO_CID_SNAPSHOT_MODE`.

```c
return video_init_ctrl(&ctrls->snapshot, dev, VIDEO_CID_SNAPSHOT_MODE,
		       (struct video_ctrl_range){.min = 0, .max = 1, .step = 1,
						 .def = data->ctrls.snapshot.val});
```

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -332,6 +434,32 @@ static int video_stm32_dcmi_get_caps(const struct device *dev, struct video_caps
+	}
+
+	return -ENOTSUP;
+}
```
I am tempted to simplify this to being just this:

```c
static int video_stm32_dcmi_set_ctrl(const struct device *dev, uint32_t id)
{
	return 0;
}
```

As the validation will already be done by the video stack.

And the value is already stored.

There is no hardware side-effect to handle *at this point*.

And `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX` is not a reliable way to know how many buffers are going to be submitted to DCMI (for instance, could be buffers for LVGL to send them to UVC).

KurtE replied:
Thanks @josuah - I need to get my head around this again.  Been trying to get some things working at the ArduinoCore-zephyr level.  - They switched a lot of the code to be deferred-init, which is good.  I had the camera working that way maybe 7-8 months ago, although never pulled in.  However then Arduino took it up a notch and also made the I2C objects to be deferred as well, and so I could not figure out how to walk up the chain to get the I2C object and initialize it before you try to initialize the camera...

But as of this morning, it looks like an Arduino developer has now gotten that to work (their PR #499), which I am testing out... But now back to it camera erroring out.
`[00:00:11.426,000] <wrn> video_stm32_dcmi: HAL_DCMI_ErrorCallback
`
Which I know some of these were fixed earlier on.  But they are running on a modified Zephyr 4.2.0....

Looks like you have a lot of god suggestions. 
Thanks

KurtE replied:
I updated int video_stm32_dcmi_set_ctrl

I know at least some stuff appears to be working, this is Portenta H7 running in  snapshot mode.
I rebased for recent sources... 

Had to update some of test apps, for things that changed. 

<img width="269" height="344" alt="image" src="https://github.com/user-attachments/assets/f19bba3b-abc7-4281-aa2c-ffee4e4038cf" />

Pushing up these changes.

josuah replied:
Thanks for collecting the changes from all sides. This will be a lot easier to aggregate all the fixes here.

josuah reviewed (commented):
This is a very late review, sorry to keep you waiting so long!

erwango commented:
@KurtE Would you mind cleaning-up your branch ?

KurtE commented:
> @KurtE Would you mind cleaning-up your branch ?

Sorry, Yikes

I need to figure out what happened, probably my github got confused between trying to update Arduino fork and this fork...

Looks like I did a merge instead of a rebase... undid and rebased.

KurtE commented:
@josuah - I believe I put in the suggested changes.
There are a few sonar... minor warnings on lines that look like all the others in the video drivers:
like:
`		k_fifo_put(&dev_data->fifo_out, vbuf);
`

Let me know if I missed anything, have done some testing, runs on Portenta H7.

rrooding commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +337,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+
+		hal_ret = HAL_DCMI_Start_DMA(&data->hdcmi, DCMI_MODE_SNAPSHOT,
+				       (uint32_t)data->vbuf->buffer,
+				       data->vbuf->size / 4);
```
I think you will need to use data->vbuf->bytesused instead of size here, otherwise if you switch from a large image format to a small one, this configures the DMA for far more segments than the sensor will ever produce in one snapshot frame and dequeue times out. At least I ran into this while trying to do the same here when capturing different resolution images after another.

KurtE replied:
Thanks, 

You might be right.  I noticed two other HAL_DCMI_START_DMA calls in this file one using the size and other using the bytesused.   Probably has not been found as an issue before, with the recovery call,  will change that one as well to see how it works.

KurtE replied:
I believe you are right and changed both of the others to bytesused.  
Verified that it is working so far on Arduino Nicla Vision
<img width="661" height="827" alt="image" src="https://github.com/user-attachments/assets/4b1f5761-422c-4d80-b124-ef018792f784" />

Pushed up the changes

rrooding commented on dts/bindings/video/st,stm32-dcmi.yaml:
```diff
@@ -29,6 +29,15 @@ properties:
+
+  snapshot-timeout:
+    type: int
+    description: Set capture timeout in microseconds, default is 1000.
```
The description says "microseconds" but the implementation compares it against k_uptime_get_32() which returns milliseconds

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -454,13 +562,24 @@ static DEVICE_API(video, video_stm32_dcmi_driver_api) = {
 	.set_selection = video_stm32_dcmi_set_selection,
 	.get_selection = video_stm32_dcmi_get_selection,
 };
 
+static int video_stm32_dcmi_init_controls(const struct device *dev)
+{
+	struct video_stm32_dcmi_data *data = dev->data;
+	struct video_stm32_dcmi_ctrls *ctrls = &data->ctrls;
+
+	return video_init_ctrl(&ctrls->snapshot, dev, VIDEO_CID_SNAPSHOT_MODE,
+			       (struct video_ctrl_range){.min = 0, .max = 1, .step = 1,
+							 .def = data->ctrls.snapshot.val});
+}
```
I think this is where you meant to put the default value:

```diff
 	return video_init_ctrl(&ctrls->snapshot, dev, VIDEO_CID_SNAPSHOT_MODE,
 			       (struct video_ctrl_range){.min = 0, .max = 1, .step = 1,
-							 .def = data->ctrls.snapshot.val});
+							 .def = cfg->snapshot_mode});
```

KurtE replied:
Yes, I meant to change that one.  Again assuming it only gets called once at startup.

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -571,6 +693,13 @@ static int video_stm32_dcmi_init(const struct device *dev)
+		data->ctrls.snapshot.val = 0;
+	} else {
+		data->ctrls.snapshot.val = config->snapshot_mode ? 1 : 0;
+	}
```
It is nice to have systems that automatically configure themselves.

Though, if the user specifies `snapshot-mode;` on devicetree, they would expect snapshot mode to be enabled.

It is possible to make this driver work with snapshot mode with `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=0`, and passing local `struct video_buffers vbuf;` variables from `main()`.

By using `config->snapshot_mode` as default control value, maybe this does what you want: default from devicetree, still possible to adjust it with video controls.

josuah replied:
Maybe this is the solution to all our "multi-camera library" problems: make every devicetree config runtime-modifiable as a control value.

KurtE replied:
>Though, if the user specifies snapshot-mode; on devicetree, they would expect snapshot mode to be enabled.

The else clause should that, if it is in the config then it sets it to 1... 

Although the if clause looks wrong... val should = 1, not 0...

KurtE replied:
> It is possible to make this driver work with snapshot mode with CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=0, and passing local struct video_buffers vbuf; variables from main().

Interesting suggestion, will have to figure out how to get it passed through... Would make sense

josuah replied:
I would be tempted to remove `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX` completely from the driver to be honest.

It just does not help in any way to know how many buffers the driver will receive.

This completely depends on what the application does and what these buffers will be used for.

For instance, there can be buffers allocated to a JPEG encoder, or a virtual driver (video-sw-generator)...

I would remove this entire if/else completely and replace it by nothing.

Instead, above, the `def = ` will set the default value for the video control, and this would have the effect of letting the devicetree select the default for snapshot/continuous mode.

KurtE replied:
I will look at that, makes sense, for the most part the driver simply works with the two queues, not with the count

KurtE replied:
I pushed up those changes, so far I have tried on Portenta H7... Having some issues with Nicla Vision, but probably something not configured correctly.  I2C Errors

josuah replied:
With Nicla vision I ended-up having the connector simply unplugged and felt silly to search what was going on for so long.

KurtE replied:
ding ding ding!
It was loose.... Aargh   Thanks

<img width="487" height="358" alt="image" src="https://github.com/user-attachments/assets/fffebebb-2aed-4e05-a0d4-aade8721d983" />

This as with 480x320 capture. using @dani-clo test python viewer
https://github.com/dani-clo/Arduino_Sketches/tree/main/CameraCaptureSerialViewer

running on my Windows 11 machine:
`D:\github\Arduino_Sketches\CameraCaptureSerialViewer>python camera_serial_viewer.py --port com13 --width 480 --height 320 --format rgb565_le`

running with this PR, with the app configured to use single buffer.

josuah replied:
Thanks for the tests!

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -270,22 +289,28 @@ static int video_stm32_dcmi_set_stream(const struct device *dev, bool enable,
-	if (data->vbuf == NULL) {
-		LOG_ERR("Failed to dequeue a DCMI buffer.");
-		return -ENOMEM;
+		if (data->vbuf == NULL) {
+			data->ctrls.snapshot.val = 1;
+			LOG_WRN("No buffers, assume snapshot mode.");
+		}
```
I have the impression if we run out of buffers during continuous mode (`data->vbuf == NULL`), it would switch to snapshot mode?

Maybe it is enough to return `-ENOMEM`?

KurtE replied:
At least how it used to work and I believe it still does....

In continuous mode, there is one buffer that is always allocated to the camera, and when it gets a frame complete, the callback function tries to allocate a buffer from the free queue, if it retrieves one, it memcpy's the received frame to this buffer and puts that buffer at the tail of the pending frame queue (I probably should look up the actual names of the queues).  If there is no free buffers on the free queue, it tosses that frame and starts up receiving the next....

Personally I don't believe this is great, as suppose you setup your queue with three buffers.  One is always on the camera, so when the first frame is received it grabs a buffer and puts it on the available queue, and the app grabs it and maybe it starts up an update of a larger TFT display that is updated over SPI, and it takes lets say 1/4 of a second to update the display, and then it returns that buffer to the free queue.  During that time frame, lf the camera is returning lets says 30 frames per second, the next frame will come in really quick and the third buffer is removed from free queue, memcpy, put on available queue, and then for the next almost 1/4 second another 8 frames are received and tossed.  Then right after the display is updated, it grabs the first (only) one from the available list, which is more or less 1/4 second old (stale), which is bad enough, now suppose instead I am using a USBToSerial to the Arducam display and it takes lets say 5 seconds to receive an image, then the next image is 5 seconds old...

There are other options on this:

a) Use snapshot - then when it finished that image, it starts the camera again, depending on where the camera is, it could take nearly two frame times to retrieve an image, that is it first waits for a start of frame indication and then it starts retaining the data.

b) Define to use 2 buffers;  It receives one, the app starts processing that frame, and during this time the camera is spinning, receiving frame after frame and tossing, until the app, hands back it's buffer, and when the next frame is completed, it copies and returns it... So waiting for next frame is anywhere from almost 0 to time to receive a full frame.  Plus it can be faster, minus: inconsistent timing, plus more things hitting memory at the same time, so if using SDRAM, you will have multiple things hitting the SDRAM at the same time, leading to more DMA errors.. (same for > 2 buffers, although not maybe not as bad, ad you won't have memcpy going on at the same time as the display touching it.

c) More than 2 buffers - instead of toss, reuse the oldest buffer in the ready list...  So if using 3 buffers, where one buffer is used by camera, other being used by app, it would update that last buffer each time a frame is received.  So when the app has completed processing an image, and hands back that buffer and grabs another, it would be retrieving the last full buffer received.

Which is best?  I am not sure one size fits all.  Some may want the last N frames, others might want the first N frames.  Others may want reduced power and memory contention...

Sorry for long winded answer.  Hope some of it makes some sense.

josuah replied:
> the next frame will come in really quick and the third buffer is removed from free queue

I remember you discussing about queuing challenges where you needed to have the most recent frame available.

Trying to summarize making sure I got it:

a) insert buffer, ask for pixels to come, pick buffer when full

b) pixels constantly refreshed on an internal buffer, present a new buffer to get a copy of it

c) insert buffers on the input queue, the next one gets refreshed until it gets picked up

> Which is best? I am not sure one size fits all.

It looks like (c) is a better (b), I see no reason to keep (b) once (c) is there.
This would require swapping the buffer during the interrupt.

(c) would further remove the problem of "first/last" frame out of the driver: frames can constantly be inserted and discarded by the application: now that there is no more `memcpy()`.

Application keeps a pointer to "most_recent_vbuf" that's always set to the latest frame.

With (b), this was necessary to carefully choosing the moment where a frame was submitted as it would trigger a costly `memcpy()`...

With (a) of course as you said there is least amount of power, and everything is synchronous (asking a frame and waiting for it, then passing it to the display) so no data transmitted/copied unless it's really needed.

=> This brings us (a) + (b or c) as the two modes? For now (a) + (b) as this PR is only about adding (a) (snapshot mode), and if interested in getting rid of the `memcpy()`, this can be done in the future.

josuah replied:
But to come back to the original idea, there's probably no need to switch between the modes automatically, users can take their own decisions, and if there is some kind of dynamic selection to be done, it's possible to add that outside of the drivers.

i.e. libraries targeting one particular usage can select the most appropriate mode for this usage. Possibly an upcoming `subsys/camera` where platform specific integration can happen on top of the video API.

KurtE replied:
I just pushed up changes removing the switch.  It now reports an ERR instead of WRN and returns -ENOMEM.

This was originally put in, before using a Control to switch modes as the only way I could switch at run time.
So it is no longer necessary.

josuah replied:
ah right, then there needed some selection mechanism, which was not bad idea but IMHO your new version is better.

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+		if (*vbuf != NULL) {
+			LOG_DBG("k_fifo_get returned: %p", *vbuf);
+			if (HAL_DCMI_Stop(&data->hdcmi) != HAL_OK) {
+				LOG_WRN("Snapshot: HAL_DCMI_Stop FAILED!");
```
What about `return -EIO` after this line?

KurtE replied:
done

avolmat-st replied:
Indeed, getting an error after HAL_DCMI_Stop doesn't sound correct behavior but here it seems to return 0 after that.

josuah commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+			return -ENOMEM;
+		}
+
+		hal_ret = HAL_DCMI_Start_DMA(&data->hdcmi, DCMI_MODE_SNAPSHOT,
+				       (uint32_t)data->vbuf->buffer,
+				       data->vbuf->bytesused / 4);
+		if (hal_ret != HAL_OK) {
+			LOG_WRN("Snapshot: HAL_DCMI_Start_DMA FAILED!");
+			return -EIO;
+		}
```
As I understand, the snapshot mode works this way:

1. User enqueues buffers on the device
2. User tries to dequeue buffers from the device
3. Buffers are sent to the hardware
4. Buffers are returned from the hardware
5. User completes the "dequeue" operation

I think in the future it could be useful to have a separate `VIDEO_CID_TRIGGER_CAPTURE` or alike.

Then it's possible to have a thread that continuously enqueues/dequeues buffers regardless of the mode (continuous/snapshot), with the trigger that can be done from anywhere (i.e. in reaction to a GPIO pin on a different thread).

Good for a future improvement though, I think you found the best compromise with the existing APIs. :)

KurtE replied:
Sounds like a good future project.

josuah reviewed (commented):
While you are modifying the driver, maybe you will be interested in these related changes. Feel free to ask if any question.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -37,22 +35,31 @@ struct stream {
 	struct k_fifo fifo_out;
 	struct video_buffer *vbuf;
 	struct stream dma;
+	uint32_t snapshot_start_time;
```
This variable seems to only be used within the ``video_stm32_dcmi_snapshot`` function so it should be local to that function and do not have to be part of the driver data strcuture.

avolmat-st replied:
Reading again, I understand now that this has to be part of the _data structure since there can be several call to dequeue for one snapshot request.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -332,6 +430,12 @@ static int video_stm32_dcmi_get_caps(const struct device *dev, struct video_caps
+{
+	return 0;
+}
+
```
Is there really a need to have that empty function ? It seems that ctrl framework can by him self store the value into the control and calling each driver set_ctrl ops is optional.

josuah replied:
my bad I forgot :)

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -588,7 +707,8 @@ static int video_stm32_dcmi_init(const struct device *dev)
 
-	return 0;
+	/* Initialize controls */
+	return video_stm32_dcmi_init_controls(dev);
```
Could you perform this init_controls prior to the LOG_DBG message otherwise we might end up with a "dcmi inited" being displayed then failure to initialize controls.

avolmat-st commented on include/zephyr/drivers/video-controls.h:
```diff
@@ -438,6 +438,12 @@ enum video_camera_orientation {
+ * 1 - enable snapshot mode
+ */
+#define VIDEO_CID_SNAPSHOT_MODE (VIDEO_CID_PRIVATE_BASE + 0)
+
```
Since this is a private control for the DCMI, it should be defined within the stm32-dcmi.h specific header file. You might want to have a look at ``include/zephyr/drivers/video/arducam_mega.h``

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -89,6 +97,17 @@ void HAL_DCMI_FrameEventCallback(DCMI_HandleTypeDef *hdcmi)
+		dev_data->vbuf = NULL;
+		vbuf->timestamp = k_uptime_get_32();
+		k_fifo_put(&dev_data->fifo_out, vbuf);
+		LOG_DBG("event SH put: %p", vbuf);
```
Debug message to be removed.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+		/* check first to see if we already have a buffer returned */
+		*vbuf = k_fifo_get(&data->fifo_out, K_NO_WAIT);
+		if (*vbuf != NULL) {
+			LOG_DBG("k_fifo_get returned: %p", *vbuf);
```
Debug message to be removed

avolmat-st replied:
Does this check have to be part of the ``if (data->vbuf == NULL)``
I am tempted to think that if there is a buffer in the output fifo, then it has to be returned.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+			return -EIO;
+		}
+		data->vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT);
+		LOG_DBG("camera buf: %p", data->vbuf);
```
Debug message to be removed

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+		*vbuf = k_fifo_get(&data->fifo_out, K_NO_WAIT);
+		if (*vbuf != NULL) {
+			LOG_DBG("k_fifo_get returned: %p", *vbuf);
+			if (HAL_DCMI_Stop(&data->hdcmi) != HAL_OK) {
```
Why in such case stopping the DCMI is necessary ? I mean in snapshot mode the DCMI is started and will get stopped as soon as a frame has been captured so why having to stop it ?
In case of timeout, while not receiving the frame, I agree it has to be stopped but in my understanding this is done below.
I could think maybe of a case transitioning from continuous to snapshot, but in such case that should properly be taken care when switching mode ?

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+			if (HAL_DCMI_Stop(&data->hdcmi) != HAL_OK) {
+				LOG_WRN("Snapshot: HAL_DCMI_Stop FAILED!");
+			}
+			/* verify it did not come in during this procuessing */
```
s/procuessing/processing

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+
+			if (data->vbuf != NULL) {
+				k_fifo_put(&data->fifo_in, data->vbuf);
+				data->vbuf = NULL;
```
If I understand it right, when no frame has been captured during the snapshot timeout period, it is given back to the application, and error code is -EAGAIN ?
If that is so, I don't think it should be -EAGAIN, maybe more -ETIMEDOUT.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +333,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+	}
+
+	*vbuf = k_fifo_get(&data->fifo_out, timeout);
+	LOG_DBG("k_fifo_get returned: %p", *vbuf);
```
Debug message to be removed.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -542,6 +658,9 @@ static const struct video_stm32_dcmi_config video_stm32_dcmi_config_0 = {
 	.sensor_dev = SOURCE_DEV(0),
+	.snapshot_mode = DT_INST_PROP(0, snapshot_mode),
+	.snapshot_timeout = DT_INST_PROP(0, snapshot_timeout),
+	DCMI_DMA_CHANNEL(0, PERIPHERAL, MEMORY)
```
Is this being used (DCMI_DMA_CHANNEL) ? Same as the new dma variable in the config structure.

KurtE commented:
@avolmat-st @josuah - I think I covered hopefully of the comments.  I just rebuilt and after slight mod in the test sketch to include the dcmi... header file, it built and ran on the arduino_nicla_vision

avolmat-st commented:
Thanks @KurtE for the update. Could you go through the comments I made in my previous review and make comments on them ? I think some are still applicable so it would help me for the review if you could hint me on what has been modified. Thanks.

KurtE commented:
> Thanks @KurtE for the update. Could you go through the comments I made in my previous review and make comments on them ? I think some are still applicable so it would help me for the review if you could hint me on what has been modified. Thanks.

I am trying to remember where your previous comments are.  Probably int he previous PR that this was split from?
The one that was reduced to error recovery 
https://github.com/zephyrproject-rtos/zephyr/pull/93797 ?

avolmat-st commented:
I am referring to the comment I made in this PR. You can see them within the conversation tab of this PR.

That is: https://github.com/zephyrproject-rtos/zephyr/pull/98004#discussion_r3373564358, https://github.com/zephyrproject-rtos/zephyr/pull/98004#discussion_r3373589891 ... etc.

KurtE commented:
> I am referring to the comment I made in this PR. You can see them within the conversation tab of this PR.
> 
> That is: [#98004 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/98004#discussion_r3373564358), [#98004 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/98004#discussion_r3373589891) ... etc.

Sorry, I thought you were maybe suggesting a previous review like maybe months ago and I did not find any.

I believe I went through all of your comments and commented, most of which was code changed and pushed up.

I did not resolve any of them as I might have done in the past, as on a previous PR, I had a reviewer insist that I was not supposed to close an issue only he should when he was satisfied.  Unclear of preferred approach.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -332,6 +428,7 @@ static int video_stm32_dcmi_get_caps(const struct device *dev, struct video_caps
 	return video_get_caps(config->sensor_dev, caps);
 }
 
+
```
Unnecessary added line.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -586,9 +696,14 @@ static int video_stm32_dcmi_init(const struct device *dev)
+	}
 
-	return 0;
+	LOG_DBG("%s inited", dev->name);
```
Nit: please keep an empty line before the last return of a function.

avolmat-st reviewed (changes requested):
I went through the modifications.

It seems to me that below 3 are still to be addressed or commented since I am not sure about the behavior. Could you clarify ?

https://github.com/zephyrproject-rtos/zephyr/pull/98004#discussion_r3367487939
https://github.com/zephyrproject-rtos/zephyr/pull/98004#discussion_r3373965734
https://github.com/zephyrproject-rtos/zephyr/pull/98004#discussion_r3373991666

Apart from that I've added 2 very very small points.

KurtE commented:
@avolmat-st - I just verified again, that currently if you call:
```
		hal_ret = HAL_DCMI_Start_DMA(&data->hdcmi, DCMI_MODE_SNAPSHOT,
				       (uint32_t)data->vbuf->buffer,
				       data->vbuf->bytesused / 4);
```
A second time without call stop, it returns hat_ret of 1

Which is HAL_ERROR:
```
typedef enum
{
  HAL_OK       = 0x00,
  HAL_ERROR    = 0x01,
  HAL_BUSY     = 0x02,
  HAL_TIMEOUT  = 0x03
} HAL_StatusTypeDef;
```

I am guessing it is in the call to: HAL_StatusTypeDef HAL_DMA_Start_IT
Probably should learn best way to print things out from HAL layer... 

Edit: Not sure yet how to print HAL layer, but was able to print out what the HAL Layer was looking at:
```
		LOG_DBG ("HAL_DMA_Start_IT state: %u\n", data->hdcmi.DMA_Handle->State);
		hal_ret = HAL_DCMI_Start_DMA(&data->hdcmi, DCMI_MODE_SNAPSHOT,
				       (uint32_t)data->vbuf->buffer,
				       data->vbuf->bytesused / 4);
		if (hal_ret != HAL_OK) {
			LOG_WRN("Snapshot: HAL_DCMI_Start_DMA FAILED(%d)!", hal_ret);
			return -EIO;
		}
```
Which called inidrectly
```
{
  HAL_StatusTypeDef status = HAL_OK;

...

  if(HAL_DMA_STATE_READY == hdma->State)
  {
...
  }
  else
  {
    /* Set the error code to busy */
    hdma->ErrorCode = HAL_DMA_ERROR_BUSY;

    /* Process unlocked */
    __HAL_UNLOCK(hdma);

    /* Return error status */
    status = HAL_ERROR;
  }
  return status;
}
```
Debug output:
```
Serial command: 30
Send BMP: 480 320
[00:00:21.516,000] <dbg> video_stm32_dcmi.video_stm32_dcmi_snapshot: dequeue snapshot: 0 18446744073709551615
[00:00:21.516,000] <dbg> video_stm32_dcmi.video_stm32_dcmi_snapshot: HAL_DMA_Start_IT state: 1

Serial command: 30
Send BMP: 480 320
[00:00:56.292,000] <dbg> video_stm32_dcmi.video_stm32_dcmi_snapshot: dequeue snapshot: 0 18446744073709551615
[00:00:56.292,000] <dbg> video_stm32_dcmi.video_stm32_dcmi_snapshot: HAL_DMA_Start_IT state: 2

[00:00:56.292,000] <wrn> video_stm32_dcmi: Snapshot: HAL_DCMI_Start_DMA FAILED(1)!
[00:00:56.292,000] <err> main: Unable to dequeue video buf
```
So the hdma->State on first call was 1 which is HAL_DMA_STATE_READY
```
typedef enum
{
  HAL_DMA_STATE_RESET             = 0x00U,  /*!< DMA not yet initialized or disabled */
  HAL_DMA_STATE_READY             = 0x01U,  /*!< DMA initialized and ready for use   */
  HAL_DMA_STATE_BUSY              = 0x02U,  /*!< DMA process is ongoing              */
  HAL_DMA_STATE_ERROR             = 0x03U,  /*!< DMA error state                     */
  HAL_DMA_STATE_ABORT             = 0x04U,  /*!< DMA Abort state                     */
}HAL_DMA_StateTypeDef;
```
The second call, Status = 2 or HAL_DMA_STATE_BUSY, which errors out.

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -308,19 +332,91 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+		if (time_since_start_time > config->snapshot_timeout) {
+			LOG_WRN("Snapshot: Timed out!");
+			if (HAL_DCMI_Stop(&data->hdcmi) != HAL_OK) {
+				LOG_WRN("Snapshot: HAL_DCMI_Stop FAILED!");
```
Should this return an error as well here ?

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -586,9 +695,15 @@ static int video_stm32_dcmi_init(const struct device *dev)
 	}
 
 	k_sleep(K_MSEC(100));
```
Please add an empty line between this k_sleep and the control initialization.

avolmat-st reviewed (changes requested):
Hi @KurtE and sorry for the delay.
I gave it a try on a STM32L4R9I_DISCO into which I plugged the OV5640.
I confirm that it works properly in both snapshot and continuous mode.

There are however few small points:
  - the ``video_stm32_dcmi_get_caps`` function is still returning
    ```min_vbuf_count``` of 2 while actually 1 is possible if
    snapshot is enabled. This should probably depend on the value
    retrieved from the DT. Without that done, it is not possible
    to capture and set the -DCONFIG_VIDEO_CAM_NUM_BUFS=1 or
    -DCONFIG_VIDEO_BUFFER_POOL_NUM_MAX=1 since the sample app will
    refuse to start due to incompatible minimum number of buffers.
  - this review comment (https://github.com/zephyrproject-rtos/zephyr/pull/98004/changes#r3402107528) about a missing
    error code return doesn't seem to be part of your patch
  - I also added another error code return point on L382
  - @etienne-lms comment (https://github.com/zephyrproject-rtos/zephyr/pull/98004/changes#r2455034002) is also not taken into account
  - the commit log might also have to be fixed since this mention that a way to start the snapshot mode is to define
    ``CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=1``. Indeed, this used to the case in previous version of this dev however I think it is no
    more done like that so if I am correct this should be removed because it is misleading.

Could you fix those so this can be approved / merged ? Thanks a lot for all your efforts and your patience.

KurtE commented:
@avolmat-st - pushed up changes for DCMI_STOP, min buffer count, blank line.

Also updated the commit message to only show two ways... remove the count=1

avolmat-st commented:
> @avolmat-st - pushed up changes for DCMI_STOP, min buffer count, blank line.
> 
> Also updated the commit message to only show two ways... remove the count=1

Thanks for the update and correction. All seems fine, however why all HAL_DCMI_Stop return value check have been removed ? I think original request from @etienne-lms was questioning if an error should be given back when getting an error during HAL_DCMI_Stop. It seems to be that all calls were correct with error checking done on this HAL_DCMI_Stop call except for one case for which after getting an error on HAL_DCMI_Stop the caller was still returning 0 (success).

KurtE commented:
> I think original request from @etienne-lms was questioning if an error should be given back when getting an error during HAL_DCMI_Stop. It seems to be that all calls were correct with error checking done on this HAL_DCMI_Stop call except for one case for which after getting an error on HAL_DCMI_Stop the caller was still returning 0 (success).

I took them out, as they really did not appear to make any difference in the operation of the code.  That is at best they were mostly just INFO or Debug.  When they errored, it was more like it was returning an indication, that it did not do anything.  Could put them back in if needed or wanted.

Thanks
Kurt

avolmat-st commented:
> > I think original request from @etienne-lms was questioning if an error should be given back when getting an error during HAL_DCMI_Stop. It seems to be that all calls were correct with error checking done on this HAL_DCMI_Stop call except for one case for which after getting an error on HAL_DCMI_Stop the caller was still returning 0 (success).
> 
> I took them out, as they really did not appear to make any difference in the operation of the code. That is at best they were mostly just INFO or Debug. When they errored, it was more like it was returning an indication, that it did not do anything. Could put them back in if needed or wanted.
> 
> Thanks Kurt

Thanks @KurtE for the clarification. I do not feel confortable in removing the ``HAL_DCMI_Stop`` return value check.
However, having given a deeper look at how ``HAL_DCMI_Stop`` is working and in which case it can return an error, this is not really critical. However this can still highlight to an issue when a use-case behavior isn't correct. Hence, my proposal is to keep the check, but maybe adapt it, making it lighter.

I also do not want to mixup those kind of cleanup with the core modifications introduced by your commit, which is supporting the snapshot mode. Here's a proposal, what do you think ?

Could you add a 1st commit BEFORE your snapshot commit, which keep the check of HAL_DCMI_Stop error value and display a ``LOG_DBG`` in case of the return value is not HAL_OK. As I mentioned above, the error cases are not critical (in the current usage of this function in the driver), however they can give hint when getting for example a weird framerate.
So I propose to check the return value and send a LOG at debug level if an error is catched, without interrupting the process of the caller function.

For example, this would give something like:

```diff
@@ -63,16 +70,14 @@ static void stm32_dcmi_process_dma_error(DCMI_HandleTypeDef *hdcmi)
        LOG_DBG("Restart DMA after Error!");
 
        /* Lets try to recover by stopping and restart */
        if (HAL_DCMI_Stop(&dev_data->hdcmi) != HAL_OK) {
-               LOG_WRN("HAL_DCMI_Stop FAILED!");
-               return;
+               LOG_DEBUG("HAL_DCMI_Stop FAILED!");
       }
 ```
 
 I propose you can update the ``HAL_DCMI_Stop`` calls in the driver prior to you modifications, make this as a first commit, and then on top of that first commit add your snapshot commit.
 
 Does this make sense ?

KurtE commented:
Thanks @avolmat-st , will try that out hopefully today

KurtE committed:
video: stm32_dcmi: HAL_DCMI_Stop failures now DBG

When HAL_DCM_Stop return an error code, their failure does not
appear to be a real failure, but instead it is more like it
simply did not have to do anything.  As such changed from reporting
it from LOG_WRN to LOG_DBG

Signed-off-by: Kurt Eckhardt <kurte@rockisland.com>

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -63,16 +70,14 @@ static void stm32_dcmi_process_dma_error(DCMI_HandleTypeDef *hdcmi)
-		LOG_DBG("HAL_DCMI_Stop FAILED!");
-		return;
-	}
+	HAL_DCMI_Stop(&dev_data->hdcmi);
```
As discussed, we should have LOG_DBG after HAL_DCMI_Stop function call.

```suggestion
	if (HAL_DCMI_Stop(&dev_data->hdcmi) != HAL_OK) {
		LOG_DBG("HAL_DCMI_Stop FAILED!");
		return;
	}
```

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -307,25 +328,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+		/* check first to see if we already have a buffer returned */
+		*vbuf = k_fifo_get(&data->fifo_out, K_NO_WAIT);
+		if (*vbuf != NULL) {
+			HAL_DCMI_Stop(&data->hdcmi);
```
```suggestion
		hal_ret = HAL_DCMI_Stop(&data->hdcmi);
		if (hal_ret != HAL_OK) {
			LOG_DBG("Failed to stop DCMI");
		}
```

avolmat-st commented on drivers/video/video_stm32_dcmi.c:
```diff
@@ -307,25 +328,92 @@ static int video_stm32_dcmi_enqueue(const struct device *dev, struct video_buffe
+		return -ETIMEDOUT;
 	}
 
+	HAL_DCMI_Stop(&data->hdcmi);
```
```suggestion
	hal_ret = HAL_DCMI_Stop(&data->hdcmi);
	if (hal_ret != HAL_OK) {
		LOG_DBG("Failed to stop DCMI");
	}
```

avolmat-st reviewed (changes requested):
Thanks @KurtE for adding that additional commit.
Seems like there were issue when putting the 2nd commit on top of it since it seems to be removing the modification introduced by the 1st commit related to the HAL_DCMI_Stop calls. Could you fix this ?
I also added comments to add same LOG_DBG calls for HAL_DCMI_Stop calls added in the 2nd commit.
Thanks.

KurtE commented:
@avolmat-st I believe I made the changes you requested.   Possibly more than you requested, as I checked every STOP ...

I pushed that up.  The PR failed in one test... Guessing at it, it was like some test script was not found.  Which in the past implied something changed which required some new file... 

So I rebased again to the current stuff.  I built test sketch, and verified it was still working:

avolmat-st reviewed (dismissed):
Thanks a lot @KurtE for the modifications. LGTM.

KurtE commented:
@josuah @ngphibang and all - let me know if there is anything else here that needs to be changed.
Thanks
Kurt

ngphibang commented on include/zephyr/drivers/video/stm32_dcmi.h:
```diff
@@ -0,0 +1,27 @@
+#endif
+
+/** @brief Switch the camera between video mode(0) and snapshot mode(1). */
+#define VIDEO_CID_SNAPSHOT_MODE (VIDEO_CID_PRIVATE_BASE + 1)
```
Sorry but I'm not convinced we need a dedicated VIDEO_CID_SNAPSHOT_MODE control for this.

Snapshot acquisition should be implemented at the _application_ level by starting the stream, capturing a frame, and then stopping the stream, no ? In that model, "snapshot" is a usage pattern rather than a camera image property, like exposure, gain, etc.

Looking through the changes, it seems all is related to buffer & DMA behavior rather than a real "snapshot" mode which could be set / get via HW registers ...

Am I missing something ?

ngphibang replied:
If we want a "per-frame control" then I think we may need something similar to the v4l2's Request API here (https://docs.kernel.org/userspace-api/media/mediactl/request-api.html) which allows a specific configuration (exposure, gain, etc.) to be applied to specific frames.

Anyways, doing this in one driver is not harmful (maybe the reason behind this is you want to save one buffer in this mode ?). So, at least, can we rename this vendor-specific CID to something like below ?

`#define VIDEO_CID_ST_SNAPSHOT_MODE (VIDEO_CID_PRIVATE_BASE + 1)`

avolmat-st replied:
> Anyways, doing this in one driver is not harmful (maybe the reason behind this is you want to save one buffer in this mode ?). So, at least, can we rename this vendor-specific CID to something like below ?
> 
> `#define VIDEO_CID_ST_SNAPSHOT_MODE (VIDEO_CID_PRIVATE_BASE + 1)`

Very good point @ngphibang thanks for noticing, indeed having VIDEO_CID_SNAPSHOT_MODE, aka something which sounds standard in a driver private control is missleading. I also think it should be renamed.

avolmat-st reviewed (dismissed):
Thanks @KurtE for the update.

ngphibang commented:
@KurtE : Could you do a rebase as currently there are conflicts ?

KurtE commented:
Rebase done, Removed includes of header files that no longer exist.  Built and tested my test sketch on 
Arduno Nicla Vision
